### ISSUE TITLE
Remove Django itself from install_requires

### ISSUE THREAD

**Issue by andybak** [carljm/django-model-utils#183]
There's an unfortunate side effect of referencing Django itself in setup.py or requirements.txt

It rather unexpectedly results in an upgrade to Django itself if I type: "pip install django-model-utils -U"

And really - your app doesn't 'require' django as much as it's an add-on for Django so not much is lost by removing it. Is anyone likely to install django-model-utils and expect it to install Django for them? 

There's a discussion on the same issue here: https://github.com/charettes/django-admin-enhancer/issues/23

**Comment by carljm** [comment]
Hi Andy, thanks for the ticket!

Model-utils certainly does require Django; there are `import django` and `from django... import ...` all over the code. If you don't have Django installed, you can't even import any module in model-utils, let alone use anything from it. I don't think there could really be a clearer case of what it means for a Python project to have a dependency. That's why it's listed.

That said, you're right that in practice most likely nobody needs the dependency listed, because it's kind of a "tail wag the dog" situation -- anyone installing model-utils almost certainly knows it works with Django and already has Django installed, so doesn't need any automatic dependency handling. And the behavior of `pip install -U` in this case is certainly irritating.

I think I'd accept a PR to remove it.

**Comment by romgar** [comment]
I'm +1 to remove the dependency.

In theory, I prefer the packages to list their own dependencies, but it's a bit different in this Django / Django-app ecosystem.

I quite liked the idea proposed by @charettes:
> I guess I should remove it from install_requires then and maybe raise an ImproperlyConfigured exception if django.VERSION >= (1, 8).

We would then test if django.VERSION < (1, 4) here.

OTOH, if some functionalities are working in previous versions than 1.4, we will prevent the users to use them. And I prefer to warn people that it *may* won't work, than forbid them to do it (display a warning at module loading?)

In conclusion, I think for now removing the Django version from the installed requirements is enough, and `django-model-utils` is already stating which Django versions this package supports.

An interesting outcome from this discussion could be: should we try to make all the django apps follow a similar pattern for these kind of situations? Update [cookiecutter-django](https://github.com/pydanny/cookiecutter-django) to specify which Django versions your package supports, and what you want to do on other versions, so that users that download your package are aware of this?

**Comment by charettes** [comment]
Like @aaugustin said [here](https://github.com/jazzband/django-debug-toolbar/issues/818) this is really a discussion about how pip should handle `install_requires`. It's the only way for packages to declare dependencies and I really don't think the correct course of action is to force third-party package maintainers to drop them for `pip --upgrade` to behave in a sane way.

**Comment by andybak** [comment]
@charettes Problem there is that leaving django in install_requires causes problems with no great benefit and removing it has little discernible ill effect. Can you think of another way to avoid people getting their Django install upgraded unexpectedly?

**Comment by charettes** [comment]
> Can you think of another way to avoid people getting their Django install upgraded unexpectedly?

Fixing `pip install package --upgrade` to only upgrade `package` and not all of its dependencies?

**Comment by andybak** [comment]
So - we agree that this would merely be a workaround. I'm arguing it's worth it as a workaround as it's trivial and has little ill effect.

**Comment by charettes** [comment]
@andybak it has the effect of not declaring Django as a dependency which is a big deal if you try to install two third-party applications with incompatible Django version requirement.

**Comment by romgar** [comment]
> Fixing pip install package --upgrade to only upgrade package and not all of its dependencies?

Don't you think it can lead to other issues, like the package you are upgrading that is broken because one on its dependencies has not been updated properly @charettes ?

**Comment by charettes** [comment]
@romgar what I meant is that pip shouldn't upgrade dependencies that are already satisfied. For example say you upgrade your `django-foo` package which depends on Django >= 1.6 to a new version which depends on Django >= 1.8. Issuing a `pip install --upgrade django-foo` shouldn't upgrade Django to 1.10 if you already have Django 1.8.14 installed.

**Comment by andybak** [comment]
@charettes I just can't think of a scenario when I want a app that I use with Django to upgrade Django. Upgrading Django is a big deal and it's something I would choose to do after much thought. When is it ever useful for this to happen as a side-effect?

All I ever want a django (app/extension/plugin/library/whatever) to do is fail loudly if it won't work with my Django version.

**Comment by aaugustin** [comment]
IMO in the current state of things the pragmatic solution is to rely on documentation for supported Django versions.

**Comment by charettes** [comment]
> I just can't think of a scenario when I want a app that I use with Django to upgrade Django. Upgrading Django is a big deal and it's something I would choose to do after much thought. When is it ever useful for this to happen as a side-effect?

Bear with me I also agree this should never happen. Other packaging systems warn, fail or prompt you for resolution when you try to upgrade a package with conflicting dependencies but pip choose to silently upgrade them.

**Comment by andybak** [comment]
@charettes I completely agree but what do we do until pip does this properly?

**Comment by charettes** [comment]
@andybak use `pip install --upgrade --upgrade-strategy="only-if-needed"` until it becomes the default behavior?

**Comment by andybak** [comment]
Wow. That's good to know.

However - have pip devs declared an intent for it to become the default? If not - I'd still stand by this change.

**Comment by romgar** [comment]
After this long discussion, I will close this issue, as the ecosystem follows this principle of keeping dependencies, and think the issue is more on  `pip` dependency management approach.
